libraries/common/array.effekt

```diff
@@ -1,90 +1,114 @@
 module array
 
 import effekt
-import option
+import exception
 import list
 
-/**
- * A mutable 0-based indexed fixed-sized array.
- */
-extern type Array[T]
+// Primitive operations:
 
 /**
- * Creates a new Array of size `size` filled with the value `default`
+ * A mutable 0-indexed fixed-sized array.
  */
-extern io def fill[T](size: Int, default: T): Array[T] =
-  ml "Array.array (${size}, ${default})"
-  default {
-    val arr = emptyArray[T](size)
-    each(0, size) { i =>
-      put(arr, i, default)
-    }
-    return arr;
-  }
-
-/**
- * Builds a new Array of size `size` from a computation `index` which gets an index
- * and returns a value that will be on that position in the resulting array
- */
-def build[T](size: Int) { index: Int => T }: Array[T] = {
-  val arr = emptyArray[T](size)
-  each(0, size) { i =>
-    put(arr, i, index(i))
-  }
-  return arr
-}
+extern type Array[T]
 
 /**
- * Converts a List `list` to an Array
+ * Violated Postcondition: Fields contain values
  */
-def fromList[T](list: List[T]): Array[T] = {
-  val listSize = list.size();
-  val arr = emptyArray(listSize);
-
-  foreachIndex(list) { (i, head) =>
-    put(arr, i, head)
-  }
-  return arr;
-}
-
-extern pure def emptyArray[T](size: Int): Array[T] =
+extern pure def unsafeNew[T](size: Int): Array[T] =
```

Comment:
To be honest, unsafeNew is not unsafe itself. Getting an element from a non-initialized array is. Right now in this PR you are testing "unsafe" implementation details, which of course is fine. However, the user-facing API should be safe. Either, we force construction to maintain the invariant that all fields are set, or we check on reading an element whether it is NULL or not.

TBH the former will almost ALWAYS force users to use `Array[Option[T]]` or a custom option type to then just initialize the array with `None` AND check on every `get`. So I'd assume the latter is much more ergonomic.